### PR TITLE
util: Spoof Ada and later GPUs as Ampere for The Great Circle

### PR DESCRIPTION
**Pull request by Saancreed**
See the discussion in our LGD channel.

### CODE REVIEW

**Review by jp7677 — APPROVED** [review]

**Comment by shelterx** [comment]
I don't get this bug tbh, nvidia offered a workaround (set __GL_13ebad=0x1) for the bad performance so I decided to test without the spoof, seems like no matter what I do, the game won't crash now. Does it only crash on first launch without the spoof or is it some setting?

**Comment by Saancreed** [comment]
From my own experience and reports of other users, and to the surprise of nobody, it crashes when attempting to enable DLSS Frame Generation. Such are the woes of being a Vulkan game.

I'm not sure if it really needs a working implementation of nvcuda for it to work but even with the best one I'm aware of being present, at some point Vulkan will just fail to present and die for reasons that are still unclear to me.

(But if your experience is different from `Unhandled exception: unimplemented function nvcuda.dll.cuInit called in 64-bit code` and `FATAL ERROR: vkQueuePresentKHR failed with error (VK_TIMEOUT)`, I'd appreciate if you let us know.)

**Comment by Saancreed** [comment]
Whoa, this is a certainly interesting and unexpected! Could you please share a Proton log made with `DXVK_NVAPI_LOG_LEVEL=trace` and `PROTON_LOG=+nvcuda`?

Even if DLFG is working (which you should be able to confirm beyond doubt using https://github.com/NVIDIAGameWorks/Streamline/blob/main/docs/ProgrammingGuideDLSS_G.md#190-dlss-fg-indicator-text if you wish), I expect that Reflex does not so input latency is likely to be tough. If only I could convince the game to try loading `NvLowLatencyVk.dll` on my system at the very least…

**Comment by jp7677** [comment]
@shelterx Additionally, do you have `PROTON_ENABLE_NGX_UPDATER=1` in your setup (from https://download.nvidia.com/XFree86/Linux-x86_64/565.77/README/ngx.html)?

**Comment by shelterx** [comment]
I don't use PROTON_ENABLE_NGX_UPDATER
I tried using that registry setting but nothing showed up.

I'll try to verify if the DLSS setting actually does something, I'm currently in an area where FPS is maxed out... it's at 144 FPS even tho' I have vsync off?!

**Comment by Saancreed** [comment]
Again, logs please, the presence or lack of logged calls to `nvofapi64` functions would also be a good indicator of anything actually working there.

**Comment by shelterx** [comment]
Now I had some time to look at it, I noticed nvngx_dlssg.dll failed to load SveSops nvcuda.dll but the FG option was still there. nvofapi64.dll wasn't loaded at all, I think the proton prefix got messed up somehow.
So DLSS-FG being enabled was false positive I think.

I did manage to recreate the crash tho',
```
92830.729:0278:trace:nvcuda:DllMain (0x747210680000, 2, (nil))
92830.729:0278:trace:nvcuda:DllMain (0x747210680000, 3, (nil))
Unhandled exception: page fault on write access to 0x0000000000000000 in 64-bit code (0x000001404930c8).

```

**Comment by Saancreed** [comment]
> I noticed nvngx_dlssg.dll failed to load SveSops nvcuda.dll

That part worked for me, you probably did something wrong. If it didn't, I wouldn't have realized that I need nvcuda to report proper device LUIDs which I implemented in the PR linked between second and third comment above.

> nvofapi64.dll wasn't loaded at all

For me it was loaded then unloaded multiple times before the game settled on a black screen.

```
176253.687:0120:02a8:trace:loaddll:build_module Loaded L"C:\\windows\\system32\\nvofapi64.dll" at 00006FFFF2F70000: native
176253.689:0120:02a8:trace:nvcuda:wine_cuInit (0)
176253.689:0120:02a8:trace:nvcuda:wine_cuDriverGetVersion (0x835ed868)
176253.691:0120:02a8:trace:nvcuda:wine_cuDeviceGetCount (0x835ed858)
176253.691:0120:02a8:trace:nvcuda:wine_cuDeviceGet (0xa4098ec0, 0)
176253.691:0120:02a8:trace:nvcuda:wine_cuDeviceGetLuid (0x835ed870, 0x835ed860, 0)
176253.692:0120:02a8:trace:nvcuda:wine_cuDeviceGetLuid Found LUID: f3030000-00000000
176253.693:0120:02a8:trace:nvcuda:wine_cuDeviceGetAttribute (Device: 0, Attribute: 114) Value: (1)
176253.693:0120:02a8:trace:nvcuda:wine_cuCtxCreate_v2 (0xa4098ec8, 0, 0)
176253.823:0120:02a8:trace:nvcuda:wine_cuStreamCreate (0xa4098ed0, 0)
176253.823:0120:02a8:trace:nvcuda:wine_cuStreamCreate (0xa4098ed8, 0)
176253.823:0120:02a8:trace:nvcuda:wine_cuCtxPopCurrent_v2 (0xa4098ec8)
176253.831:0120:02a8:trace:nvcuda:wine_cuStreamDestroy_v2 (0x7980050ad890)
176253.831:0120:02a8:trace:nvcuda:wine_cuStreamDestroy_v2 (0x7980050ad8b0)
176253.831:0120:02a8:trace:nvcuda:wine_cuCtxDestroy_v2 (0x7980046fa0c0)
176253.891:0120:02a8:trace:loaddll:free_modref Unloaded module L"C:\\windows\\system32\\nvofapi64.dll" : native
```

Anyway, because you keep ignoring my requests for providing basic debugging information, I'm not going to bother with this any further. I shall settle on "it's broken, we now have a workaround, there's nothing else to do here".

**Comment by shelterx** [comment]
TLDR; yeah, it's probably broken, unclear if it's fixable.

I did investigate it and like I said, although maybe not very clear, in my previous post, the reason why DLSS-FG showed up in game was because of a broken proton prefix where nvcuda failed to load.
If nvcuda.dll is missing or fails to load, the DLSS-FG option shows as available, as in the screenshot. But it isn't actually working even though the option visibly there and togglable.

With a new fresh proton prefix with working nvcuda.dll and having AD100 exposed, nvngx_dlssg, nvcuda and nvofapi64 loads, which then causes the game to crash. 

So it was a false positive all along... :(

**Comment by Saancreed** [comment]
Yup, thanks for the logs. You'd probably get a bit further with my LUID nvcuda PR but it doesn't seem to be enough anyway :(

**Comment by shelterx** [comment]
@Saancreed Not sure how it's supposed to behave now but with the dev branch of nvcuda, I don't see any of the DLSS dlls being loaded, nvcuda loads and game still crashes tho'.
Now log attached, dunno if it'll help...
[steam-2677660.log](https://github.com/user-attachments/files/18126557/steam-2677660.log)
